internal/clients/manager.go

```diff
@@ -183,6 +200,18 @@ func (c *AdminAPIClientsManager) adminAPIAddressNotifyLoop() {
 
 			// Notify subscribers that the clients list has been updated.
 			c.notifyGatewayClientsSubscribers()
+
+			// Log the current state of clients.
+			c.logClientsLists("notify")
+
+		case <-readinessReconciliationTicker.C:
```

Comment:
Why do we need to introduce time based synchronization? Can't we rely on signals coming from `c.discoveredAdminAPIsNotifyChan` and perform action then (segregate discovered admin APIs into not ready and ready)?

Reply:
The reason behind the time-based synchronization is that we have to verify the readiness manually. Given we now accept all the endpoints (with an exception for terminating ones) that the reconciler discovers, we wouldn't be able to properly detect the admin API status change. The example scenario:

1. Reconciler is triggered because a new endpoint was added to the endpoint slice.
2. The endpoint is not ready because the Gateway behind it hasn't received a non-empty config yet.
3. Reconciler notifies the clients' manager about the endpoint.
4. Clients' manager verifies Admin API readiness of the discovered endpoint - it might be ready at this point, but doesn't have to. If it's ready, we put it on the active clients list. If not, we put it on the pending list.
5. Gateway becomes ready to serve Admin API requests. There's no update on the endpoint slice as the readiness probe doesn't know this. Reconciler doesn't get any update.
6. Clients' manager runs a readiness reconciliation loop after max 10s and moves the client from the pending to the active list. 

This is more or less a primitive version of health checking but implemented on KIC side. By default, the gateway readiness check also runs with 10s period, so the expected maximum latency after the Gateway Admin API becomes ready and KIC picking it up is the same.

Reply:
> Gateway becomes ready to serve Admin API requests. There's no update on the endpoint slice as the readiness probe doesn't know this. Reconciler doesn't get any update.

Are you sure about this? I'm suprised. Given that `EndpointSlice` contains the addresses,  when their statuses change e.g. their addresses `condition.ready` I'd expect to get a notification 

```
addressType: IPv4
apiVersion: discovery.k8s.io/v1
endpoints:
- addresses:
  - 10.244.0.7
  conditions:
    ready: true
    serving: true
    terminating: false
  nodeName: colima-control-plane
  targetRef:
    kind: Pod
    name: proxy-kong-547bf4c85-rxtpb
    namespace: kong
    uid: f50f671c-bb0d-4490-993b-a4f748a98e4f
- addresses:
  - 10.244.0.8
  conditions:
    ready: true
    serving: true
    terminating: false
  nodeName: colima-control-plane
  targetRef:
    kind: Pod
    name: proxy-kong-547bf4c85-hrf4q
    namespace: kong
    uid: 13f72872-5a77-4f65-92d6-a55b129860df
kind: EndpointSlice
metadata:
  annotations:
    endpoints.kubernetes.io/last-change-trigger-time: "2023-06-15T13:40:16Z"
  creationTimestamp: "2023-06-15T13:39:56Z"
  generateName: kong-admin-
  generation: 5
  labels:
    endpointslice.kubernetes.io/managed-by: endpointslice-controller.k8s.io
    kubernetes.io/service-name: kong-admin
    service.kubernetes.io/headless: ""
  name: kong-admin-9lndx
  namespace: kong
  ownerReferences:
  - apiVersion: v1
    blockOwnerDeletion: true
    controller: true
    kind: Service
    name: kong-admin
    uid: dcab3d44-7b84-41ef-bf97-078a0328634a
  resourceVersion: "9000"
  uid: bb9e9105-8e08-4a1a-b2d9-42fe44bec554
ports:
- name: admin
  port: 8444
  protocol: TCP
```